File: Algorithms/CP/Octave/cp.m
#!/usr/bin/env octave
% Sudoku Solver - Constraint Propagation (CP) Implementation in GNU Octave
% Uses bitsets for candidate tracking and MRV heuristic for cell selection

global cp_iterations;

% ============================================================================
% BITSET HELPER FUNCTIONS
% ============================================================================

% Test if digit is a candidate (bit set)
function result = has_candidate(set, digit)
    result = bitand(set, bitshift(1, digit)) != 0;
endfunction

% Add digit as candidate (set bit)
function result = add_candidate(set, digit)
    result = bitor(set, bitshift(1, digit));
endfunction

% Remove digit as candidate (clear bit)
function result = remove_candidate(set, digit)
    mask = bitshift(1, digit);
    result = bitand(set, bitcmp(mask, 16));  % 16 bits for uint16
endfunction

% Count number of candidates (popcount)
function result = count_candidates(set)
    result = 0;
    for d = 1:9
        if has_candidate(set, d)
            result = result + 1;
        endif
    endfor
endfunction

% Get first candidate digit
function result = get_first_candidate(set)
    for d = 1:9
        if has_candidate(set, d)
            result = d;
            return;
        endif
    endfor
    result = 0;
endfunction

% ============================================================================
% GRID STRUCTURE
% ============================================================================

function grid = create_grid()
    grid = struct('values', zeros(9, 9), 'candidates', zeros(9, 9, 'uint16'));
endfunction

% ============================================================================
% INITIALIZATION
% ============================================================================

function grid = init_grid(puzzle)
    grid = create_grid();

    for row = 1:9
        for col = 1:9
            if puzzle(row, col) == 0
                % Empty cell: set all candidates 1-9
                grid.values(row, col) = 0;
                grid.candidates(row, col) = hex2dec('3FE');  % bits 1-9 set
            else
                % Given clue: set single value
                digit = puzzle(row, col);
                grid.values(row, col) = digit;
                grid.candidates(row, col) = bitshift(1, digit);
            endif
        endfor
    endfor
endfunction

% ============================================================================
% PEER CALCULATION
% ============================================================================

% Get all 20 peers for a cell (row, col, box) - returns [20 x 2] array
function peers = get_peers(row, col)
    peers = zeros(20, 2);
    idx = 1;

    % Same row (8 cells, excluding self)
    for c = 1:9
        if c != col
            peers(idx, :) = [row, c];
            idx = idx + 1;
        endif
    endfor

    % Same column (8 cells, excluding self)
    for r = 1:9
        if r != row
            peers(idx, :) = [r, col];
            idx = idx + 1;
        endif
    endfor

    % Same 3x3 box (4 cells, excluding self and already counted)
    box_row = floor((row - 1) / 3) * 3 + 1;
    box_col = floor((col - 1) / 3) * 3 + 1;

    for r = box_row:(box_row + 2)
        for c = box_col:(box_col + 2)
            if r != row && c != col
                peers(idx, :) = [r, c];
                idx = idx + 1;
            endif
        endfor
    endfor
endfunction

% ============================================================================
% CONSTRAINT PROPAGATION
% ============================================================================

% Eliminate a digit from a cell's candidates
function [grid, success] = eliminate(grid, row, col, digit)
    % Check if digit is already eliminated
    if ~has_candidate(grid.candidates(row, col), digit)
        success = 1;
        return;
    endif

    % Remove digit from candidates
    grid.candidates(row, col) = remove_candidate(grid.candidates(row, col), digit);

    % Check for contradiction
    remaining = count_candidates(grid.candidates(row, col));
    if remaining == 0
        success = 0;
        return;
    endif

    % If only one candidate left, assign it (singleton elimination)
    if remaining == 1 && grid.values(row, col) == 0
        last_digit = get_first_candidate(grid.candidates(row, col));
        [grid, success] = assign(grid, row, col, last_digit);
        return;
    endif

    success = 1;
endfunction

% Assign a value to a cell
function [grid, success] = assign(grid, row, col, digit)
    global cp_iterations;

    % Increment iteration counter
    cp_iterations = cp_iterations + 1;

    % Set value
    grid.values(row, col) = digit;
    grid.candidates(row, col) = bitshift(1, digit);

    % Eliminate digit from all peers
    peers = get_peers(row, col);

    for i = 1:20
        peer_row = peers(i, 1);
        peer_col = peers(i, 2);

        [grid, success] = eliminate(grid, peer_row, peer_col, digit);
        if success == 0
            return;
        endif
    endfor

    success = 1;
endfunction

% Propagate constraints until quiescence
function [grid, success] = propagate(grid)
    changed = 1;

    while changed
        changed = 0;

        % Strategy 1: Singleton elimination
        for row = 1:9
            for col = 1:9
                if grid.values(row, col) == 0
                    num_candidates = count_candidates(grid.candidates(row, col));
                    if num_candidates == 0
                        success = 0;
                        return;
                    endif
                    if num_candidates == 1
                        digit = get_first_candidate(grid.candidates(row, col));
                        [grid, success] = assign(grid, row, col, digit);
                        if success == 0
                            return;
                        endif
                        changed = 1;
                    endif
                endif
            endfor
        endfor

        % Strategy 2: Hidden singles in rows
        for row = 1:9
            for digit = 1:9
                count = 0;
                last_col = -1;
                % Check if digit is already assigned
                already_assigned = 0;
                for col = 1:9
                    if grid.values(row, col) == digit
                        already_assigned = 1;
                        break;
                    endif
                endfor
                if already_assigned
                    continue;
                endif

                % Count positions where digit is possible
                for col = 1:9
                    if has_candidate(grid.candidates(row, col), digit)
                        count = count + 1;
                        last_col = col;
                    endif
                endfor

                if count == 1
                    [grid, success] = assign(grid, row, last_col, digit);
                    if success == 0
                        return;
                    endif
                    changed = 1;
                elseif count == 0
                    success = 0;
                    return;
                endif
            endfor
        endfor

        % Strategy 3: Hidden singles in columns
        for col = 1:9
            for digit = 1:9
                count = 0;
                last_row = -1;
                % Check if digit is already assigned
                already_assigned = 0;
                for row = 1:9
                    if grid.values(row, col) == digit
                        already_assigned = 1;
                        break;
                    endif
                endfor
                if already_assigned
                    continue;
                endif

                % Count positions where digit is possible
                for row = 1:9
                    if has_candidate(grid.candidates(row, col), digit)
                        count = count + 1;
                        last_row = row;
                    endif
                endfor

                if count == 1
                    [grid, success] = assign(grid, last_row, col, digit);
                    if success == 0
                        return;
                    endif
                    changed = 1;
                elseif count == 0
                    success = 0;
                    return;
                endif
            endfor
        endfor

        % Strategy 4: Hidden singles in boxes
        for box = 0:8
            box_row = floor(box / 3) * 3 + 1;
            box_col = mod(box, 3) * 3 + 1;

            for digit = 1:9
                count = 0;
                last_r = -1;
                last_c = -1;

                % Check if digit is already assigned
                already_assigned = 0;
                for r = box_row:(box_row + 2)
                    for c = box_col:(box_col + 2)
                        if grid.values(r, c) == digit
                            already_assigned = 1;
                            break;
                        endif
                    endfor
                    if already_assigned
                        break;
                    endif
                endfor
                if already_assigned
                    continue;
                endif

                % Count positions where digit is possible
                for r = box_row:(box_row + 2)
                    for c = box_col:(box_col + 2)
                        if has_candidate(grid.candidates(r, c), digit)
                            count = count + 1;
                            last_r = r;
                            last_c = c;
                        endif
                    endfor
                endfor

                if count == 1
                    [grid, success] = assign(grid, last_r, last_c, digit);
                    if success == 0
                        return;
                    endif
                    changed = 1;
                elseif count == 0
                    success = 0;
                    return;
                endif
            endfor
        endfor
    endwhile

    success = 1;
endfunction

% ============================================================================
% SEARCH
% ============================================================================

% Find cell with Minimum Remaining Values
function [found, mrv_row, mrv_col] = find_mrv_cell(grid)
    min_candidates = 10;
    found = 0;
    mrv_row = -1;
    mrv_col = -1;

    for r = 1:9
        for c = 1:9
            if grid.values(r, c) == 0
                num_candidates = count_candidates(grid.candidates(r, c));
                if num_candidates < min_candidates
                    min_candidates = num_candidates;
                    mrv_row = r;
                    mrv_col = c;
                    found = 1;
                endif
            endif
        endfor
    endfor
endfunction

% CP Search with backtracking
function [grid, success] = cp_search(grid)
    % Check if grid is complete
    [found, mrv_row, mrv_col] = find_mrv_cell(grid);

    if found == 0
        % No empty cells - grid is complete
        success = 1;
        return;
    endif

    % Try each candidate for the MRV cell
    candidates = grid.candidates(mrv_row, mrv_col);

    for digit = 1:9
        if has_candidate(candidates, digit)
            % Save grid state for backtracking
            grid_copy = grid;

            % Try assigning this digit
            [grid, success1] = assign(grid, mrv_row, mrv_col, digit);
            if success1 == 1
                % Assignment succeeded, propagate constraints
                [grid, success2] = propagate(grid);
                if success2 == 1
                    % Propagation succeeded, recurse
                    [grid, success3] = cp_search(grid);
                    if success3 == 1
                        success = 1;
                        return;
                    endif
                endif
            endif

            % Failed - restore grid state and try next candidate
            grid = grid_copy;
        endif
    endfor

    % All candidates exhausted - dead end
    success = 0;
endfunction

% ============================================================================
% I/O FUNCTIONS
% ============================================================================

function print_puzzle(grid)
    fprintf('\nPuzzle:\n');
    for row = 1:9
        for col = 1:9
            fprintf('%d ', grid(row, col));
        endfor
        fprintf('\n');
    endfor
endfunction

% ============================================================================
% MAIN FUNCTION
% ============================================================================

function main()
    global cp_iterations;

    args = argv();
    start_time = time();

    for i = 1:length(args)
        filename = args{i};
        if length(filename) > 7 && strcmp(filename(end-6:end), '.matrix') == 1
            % Normalize path for output
            display_path = filename;
            if strncmp(filename, '/app/Matrices/', 14) == 1
                display_path = filename(6:end);
                fprintf('../%s\n', display_path);
            else
                fprintf('%s\n', filename);
            endif

            % Read puzzle
            puzzle = zeros(9, 9);
            fid = fopen(filename, 'r');
            if fid == -1
                error('Cannot open file: %s', filename);
            endif

            line_count = 0;
            while ~feof(fid)
                line = fgetl(fid);
                if ischar(line)
                    line = strtrim(line);
                    if ~isempty(line) && line(1) != '#'
                        values = sscanf(line, '%d');
                        if length(values) == 9
                            line_count = line_count + 1;
                            puzzle(line_count, :) = values';
                            if line_count == 9
                                break;
                            endif
                        endif
                    endif
                endif
            endwhile
            fclose(fid);

            % Print initial puzzle
            for row = 1:9
                for col = 1:9
                    fprintf('%d ', puzzle(row, col));
                endfor
                fprintf('\n');
            endfor
            print_puzzle(puzzle);

            % Initialize grid with puzzle
            grid = init_grid(puzzle);

            % Solve using CP
            cp_iterations = 0;
            [grid, success] = cp_search(grid);

            if success == 1
                print_puzzle(grid.values);
                fprintf('\nSolved in Iterations=%d\n\n', cp_iterations);
            else
                fprintf('No solution found\n');
            endif
        endif
    endfor

    elapsed = time() - start_time;
    fprintf('Seconds to process %.3f\n', elapsed);
endfunction

main();

File: Algorithms/CP/Forth/cp.fs
\ Constraint Propagation (CP) Sudoku Solver - Forth Implementation
\ Mechanical translation from Python reference to preserve algorithm correctness.
\
\ Algorithm: Constraint propagation with MRV (Minimum Remaining Values) heuristic
\ - Uses bitsets to track candidate values (bits 1-9)
\ - Propagates constraints: singleton elimination, hidden singles
\ - Search with MRV cell selection for efficiency

\ Global iteration counter
variable cp-iterations

\ Grid structures
\ values[row][col] - assigned values (0 = empty)
\ candidates[row][col] - possible values per cell (bitset)
create grid-values 81 cells allot
create grid-candidates 81 cells allot

\ Solution array
create solution 81 cells allot

\ Input puzzle
create puzzle 81 cells allot

\ Helper to access 2D arrays as 1D
: grid-index ( row col -- index ) 9 * + ;

: get-value ( row col -- value )
    grid-index cells grid-values + @ ;

: set-value ( value row col -- )
    grid-index cells grid-values + ! ;

: get-candidates ( row col -- candidates )
    grid-index cells grid-candidates + @ ;

: set-candidates ( candidates row col -- )
    grid-index cells grid-candidates + ! ;

: get-puzzle ( row col -- value )
    grid-index cells puzzle + @ ;

: set-puzzle ( value row col -- )
    grid-index cells puzzle + ! ;

\ Bitset operations
: has-candidate ( candidate-set digit -- flag )
    1 swap lshift and 0<> ;

: add-candidate ( candidate-set digit -- new-set )
    1 swap lshift or ;

: remove-candidate ( candidate-set digit -- new-set )
    1 swap lshift invert and ;

: count-candidates ( candidate-set -- count )
    0 swap
    10 1 do
        dup i has-candidate if swap 1+ swap then
    loop drop ;

: get-first-candidate ( candidate-set -- digit )
    10 1 do
        dup i has-candidate if drop i unloop exit then
    loop drop 0 ;

\ Forward declarations
defer eliminate
defer assign
defer propagate

\ Get peers for a cell (row, col)
\ This is complex in Forth - we'll use a different strategy
\ Instead of returning a list, we'll iterate and call a function

\ eliminate-from-peer - eliminate digit from peer cell (using locals)
: eliminate-from-peer { digit peer-row peer-col -- success }
    \ Check if digit is already eliminated
    peer-row peer-col get-candidates digit has-candidate invert if
        true exit
    then

    \ Remove digit from candidates
    peer-row peer-col get-candidates digit remove-candidate
    peer-row peer-col set-candidates

    \ Check for contradiction (no candidates left)
    peer-row peer-col get-candidates count-candidates { remaining }
    remaining 0= if
        false exit
    then

    \ Singleton elimination: if only one candidate left and cell is empty
    remaining 1 = peer-row peer-col get-value 0= and if
        peer-row peer-col get-candidates get-first-candidate { last-digit }
        last-digit peer-row peer-col assign invert if
            false exit
        then
    then

    true ;

\ Assign digit to cell at (row, col) - using locals for clarity
: do-assign { digit row col -- success }
    1 cp-iterations +!  \ Increment iteration counter

    \ Set value
    digit row col set-value

    \ Set candidates to just this digit
    1 digit lshift row col set-candidates

    \ Eliminate from peers - same row
    9 0 do
        i col <> if
            digit row i eliminate-from-peer invert if
                false unloop exit
            then
        then
    loop

    \ Eliminate from peers - same column
    9 0 do
        i row <> if
            digit i col eliminate-from-peer invert if
                false unloop exit
            then
        then
    loop

    \ Eliminate from peers - same box
    row 3 / 3 *  \ box-row-start
    col 3 / 3 *  \ box-col-start
    { box-row-start box-col-start }
    box-row-start 3 + box-row-start do
        box-col-start 3 + box-col-start do
            j row <> i col <> or if
                digit j i eliminate-from-peer invert if
                    false unloop unloop exit
                then
            then
        loop
    loop

    true ;

' do-assign is assign

\ Variables for propagate
variable prop-changed
variable prop-row
variable prop-col
variable prop-digit
variable prop-count
variable prop-last
variable prop-assigned

\ Helper: check hidden single in row
: check-hs-row ( row digit -- success changed )
    prop-digit ! prop-row !
    0 prop-count !
    -1 prop-last !
    false prop-assigned !

    9 0 do
        prop-row @ i get-value prop-digit @ = if
            true prop-assigned !
            leave
        then
        prop-row @ i get-candidates prop-digit @ has-candidate if
            prop-count @ 1+ prop-count !
            i prop-last !
        then
    loop

    prop-assigned @ if
        true false exit
    then

    prop-count @ 1 = if
        prop-digit @ prop-row @ prop-last @ assign invert if
            false false exit
        then
        true true exit
    then

    prop-count @ 0= if
        false false exit
    then

    true false ;

\ Helper: check hidden single in column
: check-hs-col ( col digit -- success changed )
    prop-digit ! prop-col !
    0 prop-count !
    -1 prop-last !
    false prop-assigned !

    9 0 do
        i prop-col @ get-value prop-digit @ = if
            true prop-assigned !
            leave
        then
        i prop-col @ get-candidates prop-digit @ has-candidate if
            prop-count @ 1+ prop-count !
            i prop-last !
        then
    loop

    prop-assigned @ if
        true false exit
    then

    prop-count @ 1 = if
        prop-digit @ prop-last @ prop-col @ assign invert if
            false false exit
        then
        true true exit
    then

    prop-count @ 0= if
        false false exit
    then

    true false ;

\ Variables for box hidden singles
variable box-row-start
variable box-col-start
variable last-br
variable last-bc

\ Helper: check hidden single in box
: check-hs-box ( box digit -- success changed )
    prop-digit ! { box }
    box 3 / 3 * box-row-start !
    box 3 mod 3 * box-col-start !
    0 prop-count !
    -1 last-br !
    -1 last-bc !
    false prop-assigned !

    box-row-start @ 3 + box-row-start @ do
        box-col-start @ 3 + box-col-start @ do
            j i get-value prop-digit @ = if
                true prop-assigned !
            then
            prop-assigned @ invert if
                j i get-candidates prop-digit @ has-candidate if
                    prop-count @ 1+ prop-count !
                    j last-br !
                    i last-bc !
                then
            then
        loop
    loop

    prop-assigned @ if
        true false exit
    then

    prop-count @ 1 = if
        prop-digit @ last-br @ last-bc @ assign invert if
            false false exit
        then
        true true exit
    then

    prop-count @ 0= if
        false false exit
    then

    true false ;

\ Propagate constraints until quiescence
: do-propagate ( -- success )
    begin
        false prop-changed !

        \ Strategy 1: Singleton elimination
        9 0 do
            9 0 do
                j i get-value 0= if
                    j i get-candidates count-candidates
                    dup 0= if drop unloop unloop false exit then
                    1 = if
                        j i get-candidates get-first-candidate j i assign invert if
                            unloop unloop false exit
                        then
                        true prop-changed !
                    then
                then
            loop
        loop

        \ Strategy 2: Hidden singles - rows
        9 0 do  \ row
            10 1 do  \ digit
                j i check-hs-row  ( success changed )
                if true prop-changed ! then
                invert if unloop unloop false exit then
            loop
        loop

        \ Strategy 2: Hidden singles - columns
        9 0 do  \ col
            10 1 do  \ digit
                j i check-hs-col  ( success changed )
                if true prop-changed ! then
                invert if unloop unloop false exit then
            loop
        loop

        \ Strategy 2: Hidden singles - boxes
        9 0 do  \ box
            10 1 do  \ digit
                j i check-hs-box  ( success changed )
                if true prop-changed ! then
                invert if unloop unloop false exit then
            loop
        loop

        prop-changed @ invert
    until
    true ;

' do-propagate is propagate

\ Find MRV cell - using locals for clarity
: find-mrv-cell ( -- row col found )
    10  \ min-candidates
    -1  \ mrv-row
    -1  \ mrv-col
    false  \ found
    { min-cand mrv-row mrv-col found }

    9 0 do
        9 0 do
            j i get-value 0= if
                j i get-candidates count-candidates { num-cand }
                num-cand min-cand < if
                    num-cand to min-cand
                    j to mrv-row
                    i to mrv-col
                    true to found
                then
            then
        loop
    loop

    mrv-row mrv-col found ;

\ State stack for recursive backtracking
\ Max depth is 81 (one per empty cell)
\ Each level stores 81 values and 81 candidates = 162 cells
81 162 * constant STATE-STACK-SIZE
create state-stack STATE-STACK-SIZE cells allot
variable state-depth

: state-level-addr ( level -- addr )
    162 * cells state-stack + ;

: save-grid ( -- )
    state-depth @ state-level-addr { base }
    81 0 do
        i cells grid-values + @
        base i cells + !
        i cells grid-candidates + @
        base i 81 + cells + !
    loop
    1 state-depth +! ;

: restore-grid ( -- )
    -1 state-depth +!
    state-depth @ state-level-addr { base }
    81 0 do
        base i cells + @
        i cells grid-values + !
        base i 81 + cells + @
        i cells grid-candidates + !
    loop ;

\ CP search with recursion - using locals for clarity
: cp-search ( -- success )
    find-mrv-cell { mrv-row mrv-col found }

    found invert if
        \ No empty cells - solved! Extract solution
        9 0 do
            9 0 do
                j i get-value
                j i grid-index cells solution + !
            loop
        loop
        true exit
    then

    \ Get candidates for MRV cell
    mrv-row mrv-col get-candidates { candidates }

    \ Try each candidate digit
    10 1 do
        candidates i has-candidate if
            save-grid
            i mrv-row mrv-col assign if
                propagate if
                    recurse if
                        true unloop exit
                    then
                then
            then
            restore-grid
        then
    loop

    false ;

\ Initialize grid from puzzle
: init-grid ( -- )
    9 0 do
        9 0 do
            j i get-puzzle 0= if
                \ Empty cell: set all candidates 1-9
                0 j i set-value
                0x3FE j i set-candidates  \ bits 1-9
            else
                \ Given clue: set single value
                j i get-puzzle dup j i set-value
                1 swap lshift j i set-candidates
            then
        loop
    loop ;

\ Print puzzle
: .puzzle ( -- )
    cr ." Puzzle:" cr
    9 0 do
        9 0 do
            j i get-value .
        loop cr
    loop ;

: .solution ( -- )
    cr ." Puzzle:" cr
    9 0 do
        9 0 do
            j i grid-index cells solution + @ .
        loop cr
    loop ;

: .puzzle-input ( -- )
    9 0 do
        9 0 do
            j i get-puzzle .
        loop cr
    loop ;

\ Read matrix file
: read-matrix { c-addr u -- }
    c-addr u r/o open-file throw { fd }
    0 0 { r c }
    begin
        pad 1 fd read-file throw 0>
    while
        pad c@
        dup [char] 0 >= over [char] 9 <= and if
            [char] 0 - r c set-puzzle
            c 1+ to c
            c 9 = if 0 to c r 1+ to r then
        else
            dup [char] . = if
                0 r c set-puzzle
                c 1+ to c
                c 9 = if 0 to c r 1+ to r then
            else
                drop
            then
        then
    repeat
    fd close-file throw ;

\ Main program
: main ( -- )
    0 cp-iterations !
    0 state-depth !

    \ Get filename from command line
    next-arg dup if
        \ Print filename
        2dup type cr

        \ Read puzzle
        2dup read-matrix
        .puzzle-input

        \ Initialize grid
        init-grid
        .puzzle

        \ Initial propagation
        propagate invert if
            cr ." No solution found (contradiction during initial propagation)" cr
            cr ." Solved in Iterations=" cp-iterations @ . cr
            bye
        then

        \ Run search
        cp-search if
            .solution
            cr ." Solved in Iterations=" cp-iterations @ . cr
        else
            cr ." No solution found" cr
        then
    else
        2drop ." Usage: gforth cp.fs <matrix_file>" cr
    then
    bye ;

main
